src/tools/rest-backend/authentication_application.cpp

```diff
@@ -0,0 +1,307 @@
+ *		  attempt succeeds, a session token will be returned, otherwise an error.
+ */
+void AuthenticationApp::authenticate ()
+{
+	RootApp::setCORSHeaders (response (), "POST,OPTIONS");
+
+	if (request ().request_method () == "POST")
+	{
+		// check if request data is of type application/json
+		if (request ().content_type_parsed ().media_type () != "application/json")
+		{
+			RootApp::setNotAcceptable (response (), "You have supplied an unsupported Content-Type.",
+						   "REQUEST_UNSUPPORTED_CONTENT_TYPE");
+			return;
+		}
+
+		// try to parse request body data
+		cppcms::json::value requestData;
+		try
+		{
+			requestData = RootApp::parsePostDataAsJson (request ());
+		}
+		catch (kdbrest::exception::InvalidPostDataFormatException & e)
+		{
+			RootApp::setBadRequest (response (), "The submitted data is not of type application/json.",
+						"REQUEST_MALFORMED_DATA");
+			return;
+		}
+
+		// required request fields
+		std::string username;
+		std::string password;
+
+		// find username
+		try
+		{
+			username = requestData.get<std::string> ("username");
+		}
+		catch (cppcms::json::bad_value_cast & e)
+		{
+			RootApp::setBadRequest (response (), "You have to supply an username.", "AUTH_MISSING_USERNAME");
+			return;
+		}
+
+		// find password
+		try
+		{
+			password = requestData.get<std::string> ("password");
+		}
+		catch (cppcms::json::bad_value_cast & e)
+		{
+			RootApp::setBadRequest (response (), "You have to supply a password.", "AUTH_MISSING_PASSWORD");
+			return;
+		}
+
+		// regex checks
+		std::regex regex_username (REGEX_USERNAME);
+
+		// before doing a lookup, validate that username input is safe
+		if (!std::regex_match (username, regex_username))
+		{
+			RootApp::setUnauthorized (response (), "An account for the given username does not exist.",
+						  "AUTH_UNKNOWN_USERNAME");
+			return;
+		}
+
+		// attempt the actual authentication
+		// first load the user
+		model::User u (std::string (""));
+		try
+		{
+			u = service::StorageEngine::instance ().getUser (username);
+		}
+		catch (exception::UserNotFoundException & e)
+		{
+			RootApp::setUnauthorized (response (), "An account for the given username does not exist.",
+						  "AUTH_UNKNOWN_USERNAME");
+			return;
+		}
+		// then compare the password
+		unsigned char sha_out[SHA256_DIGEST_LENGTH];
+		if (!crypto::sha256_encrypt (const_cast<unsigned char *> (reinterpret_cast<unsigned const char *> (password.c_str ())),
+					     strlen (password.c_str ()), sha_out))
+		{
+			RootApp::setInternalServerError (response (), "Could not hash password. Please try again.", "AUTH_UNKNOWN_ERROR");
+			return;
+		}
+		if (u.getPasswordHash ().compare (std::string (reinterpret_cast<const char *> (sha_out))) != 0)
+		{
+			RootApp::setUnauthorized (response (), "The entered password is wrong.", "AUTH_INVALID_PASSWORD");
+			return;
+		}
+
+		/* authentication was successful, so lets build the token */
+		::jwt_t * jwt;
```

Comment:
> How do I invoke jwt_new with a smart pointer (function takes pointer to a pointer)?
> As the jwt is initialized after the pointer is created, how do I ensure the deleter does not get called if it was not initialized properly?

Simply call `jwt_new` and after checking that it worked initialize the `unqiue_ptr` (raw pointer as parameter). So something like:

```c++
jwt_t * jwt;
jwt_new (jwt);
if (!jwt) { return; } // do something...
std::unique_ptr<jwt_t, void(*)(jwt_t*)> mjwt (jwt, jwt_free);
```